### PR TITLE
refactor _replace_linear_8da4w

### PR DESCRIPTION
**Pull request by Hanxian97**
Summary:
Reimplement the _replace_linear_8da4w function using the more general util function _replace_with_custom_fn_if_matches_filter from torchao.quantization.quant_api to reduce code duplication of similar logic.

Test Plan:
python test/quantization/test_quant_api.py
python test/quantization/test_quant_primitives.py

### CODE REVIEW

File: torchao/quantization/GPTQ.py [hunk 1]
```diff
@@ -893,6 +893,34 @@ def _replace_linear_8da4w(
     linear_class: Type[torch.nn.Module],
     copy_weights: bool = False,
 ):
+
+    from torchao.quantization.quant_api import _replace_with_custom_fn_if_matches_filter
```
Comment (andrewor14) [inline torchao/quantization/GPTQ.py:897]: This should be moved to the top of the file

File: torchao/quantization/GPTQ.py [hunk 2]
```diff
@@ -893,6 +893,34 @@ def _replace_linear_8da4w(
...
+                    precision=precision,
+                    scales_precision=scales_precision,
+                )
+        # In distributed training, the model may be instantiated
+        # on the meta device, in which case there is no need to
+        # copy the weights, and doing so will result in an error
+        if copy_weights and child.weight.device != torch.device("meta"):
+            new_linear.weight = child.weight
+        return new_linear
+        #setattr(module, name, new_linear)
```
Comment (andrewor14) [inline torchao/quantization/GPTQ.py:918]: Please delete the commented out code (this one and the block comment in 923)

**Review by andrewor14 — COMMENTED** [review]
Thanks, looks great overall, just a few small nits!

**Comment by jerryzh168** [comment]
> _ No description provided. _

can you add a summary to talk about the context for the change?

**Comment by jerryzh168** [comment]
please make sure to fill in Summary and Test Plan: https://github.com/pytorch/ao/pull/389

**Comment by Hanxian97** [comment]
> please make sure to fill in Summary and Test Plan, like: #389

**Comment by Hanxian97** [comment]
Summary:
Reimplement the _replace_linear_8da4w function using the more general util function _replace_with_custom_fn_if_matches_filter from torchao.quantization.quant_api to reduce code duplication of similar logic.

Test Plan:
python test/quantization/test_quant_api.py
python test/quantization/test_quant_primitives.py

File: torchao/quantization/GPTQ.py (hunk 2 above)
Comment (Hanxian97) [reply torchao/quantization/GPTQ.py]: removed

File: torchao/quantization/GPTQ.py (hunk 1 above)
Comment (Hanxian97) [reply torchao/quantization/GPTQ.py:898]: added a comment here "import the util function here to avoid circular dependency"

**Comment by jerryzh168** [comment]
> Summary: Reimplement the _replace_linear_8da4w function using the more general util function _replace_with_custom_fn_if_matches_filter from torchao.quantization.quant_api to reduce code duplication of similar logic.
> 
> Test Plan: python test/quantization/test_quant_api.py python test/quantization/test_quant_primitives.py

you can edit the first message btw: 
Screenshot 2024-06-27 at 17 29 46

**Review by andrewor14 — APPROVED** [review]
Looks good to me! @jerryzh168 any other comments?

File: torchao/quantization/GPTQ.py [hunk 3]
```diff
@@ -893,10 +893,15 @@ def _replace_linear_8da4w(
     linear_class: Type[torch.nn.Module],
     copy_weights: bool = False,
 ):
-    for name, child in module.named_children():
-        if isinstance(child, nn.Linear):
-            if _check_linear_int4_k(child.in_features, groupsize) or padding_allowed:
-                new_linear = linear_class(
+    
+    #import the util function here to avoid circular dependency
```
Comment (andrewor14) [inline torchao/quantization/GPTQ.py:897]: nit: please add a space between # and import

File: torchao/quantization/GPTQ.py [hunk 4]
```diff
@@ -893,10 +893,15 @@ def _replace_linear_8da4w(
...
+    from torchao.quantization.quant_api import _replace_with_custom_fn_if_matches_filter
+
+    def filter_fn(child: torch.nn.Module, cur_fqn:str) -> bool:
```
Comment (andrewor14) [inline torchao/quantization/GPTQ.py:900]: nit: spacing `cur_fqn: str`

**Review by jerryzh168 — APPROVED** [review]
LGTM, thanks for addressing all the comments

**Comment by msaroufim** [comment]
@huydhn @clee2000 @jerryzh168 is it deliberate that we can no longer merge PRs manually now? Was this to enable ghstack? I'm not sure the tradeoff was worth it

image

**Comment by jerryzh168** [comment]
I'm not sure what happened, ghstack changes are not going to affect normal landing I think, this is unexpected. @huydhn do you know what is happening here?

**Pull request merged**